talpid-wireguard/src/lib.rs

```diff
@@ -204,48 +205,63 @@ async fn maybe_create_obfuscator(
+                    // TODO: Temporary since we may different entry IPs later?
+                    wireguard_endpoint: if config.entry_peer.endpoint.is_ipv4() {
+                        SocketAddr::from((Ipv4Addr::LOCALHOST, 51820))
+                    } else {
+                        SocketAddr::from((Ipv6Addr::LOCALHOST, 51820))
+                    },
+                    // wireguard_endpoint: config.entry_peer.endpoint,
+                    #[cfg(target_os = "linux")]
+                    fwmark: config.fwmark,
+                })
+            }
+        };
 
-                log::trace!("Patching first WireGuard peer to become {:?}", endpoint);
-                config.entry_peer.endpoint = endpoint;
+        log::trace!("Obfuscation settings: {settings:?}");
 
-                #[cfg(target_os = "android")]
-                let remote_socket_fd = obfuscator.remote_socket_fd();
+        let obfuscator = create_obfuscator(&settings)
+            .await
+            .map_err(Error::CreateObfuscatorError)?;
+        let endpoint = obfuscator.endpoint();
 
-                let (runner, abort_handle) = abortable(async move {
-                    match obfuscator.run().await {
-                        Ok(_) => {
-                            let _ = close_msg_sender.send(CloseMsg::ObfuscatorExpired);
-                        }
-                        Err(error) => {
-                            log::error!(
-                                "{}",
-                                error.display_chain_with_msg("Obfuscation controller failed")
-                            );
-                            let _ = close_msg_sender
-                                .send(CloseMsg::ObfuscatorFailed(Error::ObfuscatorError(error)));
-                        }
-                    }
-                });
-                tokio::spawn(runner);
-                return Ok(Some(ObfuscatorHandle::new(
-                    abort_handle,
-                    #[cfg(target_os = "android")]
-                    remote_socket_fd,
-                )));
+        log::trace!("Patching first WireGuard peer to become {endpoint}");
+        config.entry_peer.endpoint = endpoint;
+
+        #[cfg(target_os = "android")]
+        let remote_socket_fd = obfuscator.remote_socket_fd();
+
+        let obfuscation_task = tokio::spawn(async move {
+            match obfuscator.run().await {
+                Ok(_) => {
+                    let _ = close_msg_sender.send(CloseMsg::ObfuscatorExpired);
+                }
+                Err(error) => {
+                    log::error!(
+                        "{}",
+                        error.display_chain_with_msg("Obfuscation controller failed")
+                    );
+                    let _ = close_msg_sender
+                        .send(CloseMsg::ObfuscatorFailed(Error::ObfuscatorError(error)));
+                }
             }
-        }
+        });
+        return Ok(Some(ObfuscatorHandle::new(
```

Comment:
Undrelated to the PR, but what do you think about removing the `close_msg_sender`, returning `CloseMsg` from the join handle and awaiting that in `WireguardMonitor::wait`?

Reply:
Sounds like a great idea. But probably not for this PR.

Reply:
Alright